gamemodes/terrortown/gamemode/server/sv_player.lua

```diff
@@ -511,88 +514,76 @@ function GM:KeyPress(ply, key)
+    if ent.player_ragdoll then
+        if not ply:KeyDown(IN_WALK) then
+            CORPSE.ShowSearch(ply, ent)
         else
-            -- do nothing, can't +use held objects
-            return true
+            ply:Spectate(OBS_MODE_IN_EYE)
+            ply:SpectateEntity(ent)
         end
-    elseif tr.Entity.player_ragdoll then
-        CORPSE.ShowSearch(ply, tr.Entity, ply:KeyDown(IN_WALK) or ply:KeyDownLast(IN_WALK)) -- Body Corpse Search Identify TODO
-
-        return true
+    elseif ent:IsPlayer() and ent:IsActive() then
+        ply:Spectate(ply.spec_mode or OBS_MODE_IN_EYE)
+        ply:SpectateEntity(ent)
+    else
+        PROPSPEC.Target(ply, ent)
     end
 end
 
----
--- Normally all dead players are blocked from IN_USE on the server, meaning we
--- can't let them search bodies. This sucks because searching bodies is
--- fun. Hence on the client we override +use for specs and use this instead.
-local function SpecUseKey(ply, cmd, arg)
-    if not IsValid(ply) or not ply:IsSpec() then
+net.Receive("TTT2PlayerUseEntity", function(len, ply)
+    local ent = net.ReadEntity()
+
+    if not IsValid(ent) then
         return
     end
 
-    -- Do not allow the spectator to gather information if they're about to revive.
-    if ply:IsReviving() then
-        LANG.Msg(ply, "spec_about_to_revive", nil, MSG_MSTACK_WARN)
-
+    -- Check if the use interaction is possible
+    local distance = ply:GetShootPos():Distance(ent:GetPos())
+    if distance > 150 then
         return
     end
 
-    -- longer range than normal use
-    local tr = util.QuickTrace(ply:GetShootPos(), ply:GetAimVector() * 128, ply)
+    -- Override player use to enable addons to still handle it
+    -- Otherwise default to old IsTerror Check
+    overrideDoPlayerUse = ply:IsTerror()
+    if hook.Run("PlayerUse", ply, ent) then
+        ent:Use(ply, ply)
+    end
+    overrideDoPlayerUse = false
 
-    if not tr.Hit or not IsValid(tr.Entity) then
+    if ply:IsSpec() then
+        SpecUseKey(ply, ent)
         return
     end
 
-    if tr.Entity.player_ragdoll then
-        if not ply:KeyDown(IN_WALK) then
-            CORPSE.ShowSearch(ply, tr.Entity)
-        else
-            ply:Spectate(OBS_MODE_IN_EYE)
-            ply:SpectateEntity(tr.Entity)
+    if not ply:IsTerror() then
+        return
+    end
+
+    if ent.CanUseKey and ent.UseOverride then
+        local phys = ent:GetPhysicsObject()
+
+        if not IsValid(phys) or phys:HasGameFlag(FVPHYSICS_PLAYER_HELD) then
+            return
         end
-    elseif tr.Entity:IsPlayer() and tr.Entity:IsActive() then
-        ply:Spectate(ply.spec_mode or OBS_MODE_IN_EYE)
-        ply:SpectateEntity(tr.Entity)
-    else
-        PROPSPEC.Target(ply, tr.Entity)
+
+        ent:UseOverride(ply)
+    elseif ent.player_ragdoll then
+        CORPSE.ShowSearch(ply, ent, ply:KeyDown(IN_WALK) or ply:KeyDownLast(IN_WALK)) -- Body Corpse Search Identify TODO
+    elseif ent:IsWeapon() and ply:Alive() then
```

Comment:
just FYI, even a spectator is "alive", therefore we use `IsTerror`

Reply:
Isterror is checked some lines above.

Reply:
I'm not sure why this is added there at all. If a player `IsTerror` he is also always alive, because IsTerror is a subset of Alive. So I guess you could remove that part completely.

Reply:
IsTerror is just a team check, while IsAlive is about players living. So while it might not be possible in our gameloop to be alive and not from team terror, the checks themselves dont make sure of that. So I think it should stay for safety.

Reply:
I wrote that code in question here and I'm pretty sure I did it that way because I didn't know anything about IsTerror. If IsTerror and Alive diverge, the whole gamemode is broken anyway

Reply:
Ahh well, `Alive` doesn't make any sense here. Like I said a spectator is alive is as well and they **shouldn't** be able to pick up weapons! This is not a problem because of the previously mentioned `IsTerror` check, but that just underlines that the current `Alive` check is pretty pointless.

So while this is not really part of your PR, I'd vote in favor of removing that check to clean up the code and to avoid future confusion.

Reply:
I can do this. I dont know anything about those checks.
I see we have those terror and spec checks, that differ.
But in many parts of our codebase alive is always checked as well.
Might be, that is was just not knowing better. But as I dont know better, I would leave it in unless you are all sure its safe to remove. I dont really want to test it....

Reply:
When I started with TTT2 I used the alive checks as well as I thought that was the way to go, but it isn't. Probably something we should clean up as well.